ACE/ace/Malloc_T.cpp

```diff
@@ -797,18 +797,20 @@ ACE_Malloc_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::shared_free (void *ap)
     {
       for (;
            blockp <= currp
-             || blockp >= (MALLOC_HEADER *) currp->next_block_;
+             || blockp >= currp->next_block_;
            currp = currp->next_block_)
         {
-          if (currp >= (MALLOC_HEADER *) currp->next_block_
+          if (currp >= currp->next_block_
               && (blockp > currp
-                  || blockp < (MALLOC_HEADER *) currp->next_block_))
+                  || blockp < currp->next_block_))
             // Freed block at the start or the end of the memory pool.
             break;
         }
 
       // Join to upper neighbor.
-      if ((blockp + blockp->size_) == currp->next_block_)
+      typename ACE_CB::MALLOC_HEADER_PTR next = blockp;
+      next += static_cast<int> (blockp->size_); // ACE_Based_Pointer has += but not +
```

Comment:
Shouldn't we a type that is 64bit on 64bit systems?

Reply:
The signature in Based_Pointer_T is `void operator+= (int index);`.  An individual block would be limited to 2GiB.

Reply:
Ok